flink-end-to-end-tests/flink-batch-sql-test/src/test/java/org/apache/flink/sql/tests/Generator.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.sql.tests;
+
+import org.apache.flink.connector.testframe.source.FromElementsSource;
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.data.StringData;
+import org.apache.flink.table.data.TimestampData;
+import org.apache.flink.types.Row;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+class Generator implements FromElementsSource.ElementsSupplier<RowData> {
+    private static final long serialVersionUID = -8455653458083514261L;
+    private final List<Row> elements;
+
+    static Generator create(
+            int numKeys, float rowsPerKeyAndSecond, int durationSeconds, int offsetSeconds) {
+        int sleepMs = (int) (1000 / rowsPerKeyAndSecond);
+        InternalGenerator gen =
+                new InternalGenerator(
+                        numKeys, durationSeconds * 1000L, sleepMs, offsetSeconds * 2000L);
+        List<Row> elements = new ArrayList<>();
+        gen.forEachRemaining(elements::add);
+        return new Generator(elements);
+    }
```

Comment:
```suggestion
    static Generator create(
            int numKeys, float rowsPerKeyAndSecond, int durationSeconds, int offsetSeconds) {
        final int stepMs = (int) (1000 / rowsPerKeyAndSecond);
        final long durationMs = durationSeconds * 1000L;
        final long offsetMs = offsetSeconds * 2000L;
        final List<Row> elements = new ArrayList<>();

        int keyIndex = 0;
        long ms = 0;
        while (ms < durationMs) {
            elements.add(createRow(keyIndex++, ms, offsetMs));
            if (keyIndex >= numKeys) {
                keyIndex = 0;
                ms += stepMs;
            }
        }

        return new Generator(elements);
    }

    private static Row createRow(int keyIndex, long milliseconds, long offsetMillis) {
        return Row.of(
                keyIndex,
                LocalDateTime.ofInstant(
                        Instant.ofEpochMilli(milliseconds + offsetMillis), ZoneOffset.UTC),
                "Some payload...");
    }
```
nit: what we could also do is to get rid of the `InternalGenerator` class. It's just a while loop in the end. WDYT?

Reply:
Did not want to edit too much the original logic of the generator itself 👍 

But it makes sense 👍

Reply:
I applied your suggestion as I like it 👍

Reply:
You're right - I got a bit carried away. We shouldn't optimize it any further. But let's keep the changes as they are now since you applied it. 👍